File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Call/CallHelper/SessionTimer.m
//
//  WaitSession.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 26.05.2020.
//  Copyright © 2020 QuickBlox Team. All rights reserved.
//

#import "SessionTimer.h"

@interface SessionTimer ()

@property (nonatomic, assign) SessionTimerType type;
@property (nonatomic, strong) NSString *sessionId;
@property (nonatomic, strong) NSTimer *timer;

@end

@implementation SessionTimer

+ (id)waitSession:(NSString *)sessionId
             type:(SessionTimerType)type
      waitingTime:(NSTimeInterval)time {
    return [[SessionTimer alloc] initWithSession:sessionId
                                            type:type
                                     waitingTime:time];
}

- (id)initWithSession:(NSString *)sessionId
                 type:(SessionTimerType)type
          waitingTime:(NSTimeInterval)time {
    self = [super init];
    if (self) {
        self.type = type;
        self.sessionId = sessionId;
        self.timer = [NSTimer scheduledTimerWithTimeInterval:time
          target:self
        selector:@selector(didEndWaiting:)
        userInfo:nil
         repeats:NO];
    }
    
    return self;
}

- (void)invalidate {
    [self.timer invalidate];
}

//MARK: Internal
- (void)didEndWaiting:(NSTimer *)sender {
    if ([self.delegate respondsToSelector:@selector(timerDidEndWaiting:)]) {
        [self.delegate timerDidEndWaiting:self];
    }
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Calls/CallHelper/SessionTimer.swift
//
//  SessionTimer.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 31.03.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import UIKit

enum SessionTimerType {
    case active
    case accept
    case actions
}

protocol SessionTimerDelegate: AnyObject {
    func timerDidEndWaiting(_ timer: SessionTimer)
}

class SessionTimer: NSObject {
    //MARK: - Properties
    weak var delegate: SessionTimerDelegate?
    private(set) var type: SessionTimerType?
    private(set) var sessionId: String!
    var userInfo: [String: String]?
    private var timer: Timer?

    class func waitSession(sessionId: String, type: SessionTimerType, waitingTime time: TimeInterval) -> SessionTimer {
        return SessionTimer(sessionId: sessionId, type: type, waitingTime: time)
    }
    
    //MARK: - Life Cycle
    init(sessionId: String, type: SessionTimerType, waitingTime time: TimeInterval) {
        super.init()
        self.sessionId = sessionId
        self.type = type
        self.timer = Timer.scheduledTimer(timeInterval: time,
                                          target: self,
                                          selector: #selector(didEndWaiting(_:)),
                                          userInfo: nil,
                                          repeats: false)
    }

    //MARK: - Public Methods
    func invalidate() {
        timer?.invalidate()
    }
    
    //MARK: - Actions
    @objc func didEndWaiting(_ sender: Timer?) {
        delegate?.timerDidEndWaiting(self)      
    }
}
